In [2]:
from google.colab import files
uploaded = files.upload()

Saving Twitter- datasets.csv to Twitter- datasets.csv


Loading Dataset

In [3]:
import pandas as pd

df = pd.read_csv("Twitter- datasets.csv")
df.head()

,id,user_posted,name,description,date_posted,photos,url,tagged_users,replies,reposts,...,posts_count,profile_image_link,following,is_verified,quotes,bookmarks,parent_post_details,external_image_urls,videos,quoted_post
0,1868428607451799983,Glo███ews███,Glo███ews███,"Com o fim da ditadura Assad, muitos sírios con...","""2024-12-15T22:51:08.000Z""",NaN,https://x.com/GloboNews/status/186842860745179...,NaN,2,1,...,222223,https://pbs.twimg.com/profile_images/155910271...,122,False,1,1,"{""post_id"":null,""profile_id"":null,""profile_nam...",NaN,"[{""duration"":148167,""video_url"":""https://video...","{""data_posted"":null,""description"":null,""photos..."
1,1868159094567121215,bil███ard███,bil███ard███,Brian Austin Green Tells MGK to ‘Grow Up’ Afte...,"""2024-12-15T05:00:11.000Z""",NaN,https://x.com/billboard/status/186815909456712...,NaN,7,3,...,357584,https://pbs.twimg.com/profile_images/169657720...,3784,False,1,2,"{""post_id"":null,""profile_id"":null,""profile_nam...","[""https://pbs.twimg.com/card_img/1867636129563...",NaN,"{""data_posted"":null,""description"":null,""photos..."
2,1868451534708883739,TNT███rts███,TNT███ort███R,VENCE O PSG NO CLÁSSICO! 💪🇫🇷 Nossa @claalbuque...,"""2024-12-16T00:22:14.000Z""",NaN,https://x.com/TNTSportsBR/status/1868451534708...,"[{""biography"":null,""followers"":null,""following...",2,1,...,456734,https://pbs.twimg.com/profile_images/180701304...,859,False,0,1,"{""post_id"":null,""profile_id"":null,""profile_nam...",NaN,"[{""duration"":94861,""video_url"":""https://video....","{""data_posted"":null,""description"":null,""photos..."
3,1868441382022717466,TNT███rts███,TNT███ort███R,ÍDOLO E AGORA PRESIDENTE! 🇦🇷🇦🇷 O ex-atacante D...,"""2024-12-15T23:41:54.000Z""","[""https://pbs.twimg.com/media/Ge4JhOgXcAAQy5K....",https://x.com/TNTSportsBR/status/1868441382022...,"[{""biography"":null,""followers"":null,""following...",6,5,...,456734,https://pbs.twimg.com/profile_images/180701304...,859,False,1,3,"{""post_id"":null,""profile_id"":null,""profile_nam...",NaN,NaN,"{""data_posted"":null,""description"":null,""photos..."
4,1868418260892565925,Glo███ews███,Glo███ews███,.@DanielaLima_ : cirurgia de Lula travou negoc...,"""2024-12-15T22:10:01.000Z""",NaN,https://x.com/GloboNews/status/186841826089256...,"[{""biography"":null,""followers"":null,""following...",50,4,...,222223,https://pbs.twimg.com/profile_images/155910271...,122,False,3,3,"{""post_id"":null,""profile_id"":null,""profile_nam...",NaN,"[{""duration"":127067,""video_url"":""https://video...","{""data_posted"":null,""description"":null,""photos..."


Checking Dataset Columns

In [7]:
df.columns

Index(['id', 'user_posted', 'name', 'description', 'date_posted', 'photos',
       'url', 'tagged_users', 'replies', 'reposts', 'likes', 'views',
       'external_url', 'hashtags', 'followers', 'biography', 'posts_count',
       'profile_image_link', 'following', 'is_verified', 'quotes', 'bookmarks',
       'parent_post_details', 'external_image_urls', 'videos', 'quoted_post'],
      dtype='object')

In [8]:
posts = df["description"].dropna().tolist()

len(posts)

1000

Convert text to TF-IDF features

In [9]:
from sklearn.feature_extraction.text import TfidfVectorizer

vectorizer = TfidfVectorizer(max_features=5000, ngram_range=(1,2))
X = vectorizer.fit_transform(posts)

K-Means Clustering

In [10]:
from sklearn.cluster import KMeans

kmeans = KMeans(n_clusters=10, random_state=42)
clusters = kmeans.fit_predict(X)

print(clusters[:20])

[7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7]


CLuster Labels to dataset

In [11]:
df["cluster"] = clusters
df[["description","cluster"]].head()

,description,cluster
0,"Com o fim da ditadura Assad, muitos sírios con...",7
1,Brian Austin Green Tells MGK to ‘Grow Up’ Afte...,7
2,VENCE O PSG NO CLÁSSICO! 💪🇫🇷 Nossa @claalbuque...,7
3,ÍDOLO E AGORA PRESIDENTE! 🇦🇷🇦🇷 O ex-atacante D...,7
4,.@DanielaLima_ : cirurgia de Lula travou negoc...,7


Calculate clustering metrics

In [12]:
from sklearn.metrics import silhouette_score, calinski_harabasz_score, davies_bouldin_score

sil = silhouette_score(X, clusters)
ch = calinski_harabasz_score(X.toarray(), clusters)
db = davies_bouldin_score(X.toarray(), clusters)

print("Silhouette Score:", sil)
print("Calinski-Harabasz Index:", ch)
print("Davies-Bouldin Index:", db)

Silhouette Score: -0.015702170472076327
Calinski-Harabasz Index: 2.914511127256975
Davies-Bouldin Index: 1.6138210569762603


different cluster numbers

In [13]:
for k in range(5, 15):
    model = KMeans(n_clusters=k, random_state=42)
    labels = model.fit_predict(X)

    sil = silhouette_score(X, labels)
    print("Clusters:", k, "Silhouette Score:", sil)

Clusters: 5 Silhouette Score: 0.013150035746117658
Clusters: 6 Silhouette Score: 0.01522961816423593
Clusters: 7 Silhouette Score: 0.015081206705109814
Clusters: 8 Silhouette Score: -0.02006555888181649
Clusters: 9 Silhouette Score: -0.02114614845266865
Clusters: 10 Silhouette Score: -0.015702170472076327
Clusters: 11 Silhouette Score: -0.016534893774887024
Clusters: 12 Silhouette Score: -0.011713417051494205
Clusters: 13 Silhouette Score: -0.012637497090877841
Clusters: 14 Silhouette Score: -0.013627087466505615


Final K-Means Clustering Model (6 Clusters)

In [14]:
from sklearn.cluster import KMeans

kmeans = KMeans(n_clusters=6, random_state=42)
clusters = kmeans.fit_predict(X)

df["cluster"] = clusters
df[["description","cluster"]].head()

,description,cluster
0,"Com o fim da ditadura Assad, muitos sírios con...",5
1,Brian Austin Green Tells MGK to ‘Grow Up’ Afte...,2
2,VENCE O PSG NO CLÁSSICO! 💪🇫🇷 Nossa @claalbuque...,5
3,ÍDOLO E AGORA PRESIDENTE! 🇦🇷🇦🇷 O ex-atacante D...,5
4,.@DanielaLima_ : cirurgia de Lula travou negoc...,5


Calculating metrics again

In [15]:
from sklearn.metrics import silhouette_score, calinski_harabasz_score, davies_bouldin_score

sil = silhouette_score(X, clusters)
ch = calinski_harabasz_score(X.toarray(), clusters)
db = davies_bouldin_score(X.toarray(), clusters)

print("Final Silhouette Score:", sil)
print("Final Calinski-Harabasz Index:", ch)
print("Final Davies-Bouldin Index:", db)

Final Silhouette Score: 0.01522961816423593
Final Calinski-Harabasz Index: 7.401102288276827
Final Davies-Bouldin Index: 7.448678592209196


**Final Results**

Silhouette Score
0.0152

Calinski-Harabasz Index
7.40

Davies-Bouldin Index
7.45